### PR TITLE
Java8 -> Java17 (Excluding Tests)

### PR DESCRIPTION
**Pull request by taswartz**
Currently missing:

- Context for "main/*" endpoints and "custom/*" endpoints **This should be fixed now**
- Security **Implementing Spring security?**
- What to do with webapp folder that contains `app.config` in the .ebextensions folder and ssl config **Should not be needed once app has been containerized**
- Add in either log4j2 loggers or use `@Slf4j` annotation to bring in Lombok's logger **DONE**
- Logentries config **DONE**

### CODE REVIEW

File: src/main/resources/application.properties [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+## Hibernate configs
+hibernate.showSql=true
+hibernate.generateDdl=true
+
+logEntries.token=
+
+orca.username=
+orca.password=
+
+rest.prefix.core=/main/resttrigger
```
Comment (taswartz) [inline src/main/resources/application.properties:67]: This is going to be used as the prefix for the rest trigger endpoints. Configured in the `@Controller` annotation. I believe this replaces everything we were doing before with the main-servlet.xml and custom-servlet.xml file

File: src/main/java/com/client/config/ScheduledEventConfig.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
+package com.client.config;
```
Comment (taswartz) [inline src/main/java/com/client/config/ScheduledEventConfig.java:1]: What if we had a folder called "subscribedevents" (Name does not matter). In each folder, you create your four java beans for the scheduled task that has a subscription associated with it. Then there would be one scheduled task configuration for each task that is set up?

File: src/main/java/com/client/core/picker/controller/AbstractPickerController.java [hunk 3]
```diff
@@ -0,0 +1,76 @@
+package com.client.core.picker.controller;
```
Comment (taswartz) [inline src/main/java/com/client/core/picker/controller/AbstractPickerController.java:1]: I think we should just get rid of the entire Picker folder

File: src/main/java/com/client/config/ScheduledEventConfig.java (hunk 2 above)
Comment (scriptom) [reply src/main/java/com/client/config/ScheduledEventConfig.java:1]: Discussed over slack and PR 1 in your fork has a possible solution to this.

File: src/main/java/com/client/core/base/controller/AbstractTriggerController.java [hunk 4]
```diff
@@ -1,34 +1,27 @@
...
+import org.springframework.web.bind.WebDataBinder;
+import org.springframework.web.bind.annotation.InitBinder;
 
-public abstract class AbstractTriggerController<E extends BullhornEntity, H extends TriggerHelper<E>, T extends TriggerTraverser<E, H>>{
-
-    protected final BullhornData bullhornData;
+import java.math.BigDecimal;
 
-    protected final Logger log = Logger.getLogger(getClass());
+public abstract class AbstractTriggerController<E extends BullhornEntity, H extends TriggerHelper<E>, T extends TriggerTraverser<E, H>>{
```
Comment (scriptom) [inline src/main/java/com/client/core/base/controller/AbstractTriggerController.java:16]: Formatting - space before opening curly brace

File: src/main/java/com/client/core/base/dao/impl/StandardJpaDao.java [hunk 5]
```diff
@@ -201,9 +192,9 @@ private Long getTotal(TypedQuery<T> query) {
         query.getParameters().stream().forEach( parameter -> {
             String name = parameter.getName();
 
-            for(int position : parameterTranslations.getNamedParameterSqlLocations(name)) {
+            for(int position : parameterTranslations.getPositionalParameterInformationMap().keySet()) { // TODO: Find replacement (This could work, but needs testing)
```
Comment (scriptom) [inline src/main/java/com/client/core/base/dao/impl/StandardJpaDao.java:195]: Kinda worried about this one - changing named parameters for positional parameters kind of a big deal.

File: src/main/java/com/client/core/base/model/helper/AbstractHelper.java [hunk 6]
```diff
@@ -22,7 +22,8 @@
 
     private final Logger log = Logger.getLogger(getClass());
 
-    private final BullhornData bullhornData;
+    @Autowired
+    private BullhornData bullhornData;
```
Comment (scriptom) [inline src/main/java/com/client/core/base/model/helper/AbstractHelper.java:26]: Isn't it possible for us to inject this instead of autowiring it?

File: src/main/java/com/client/core/base/tools/entitychanger/impl/StandardEntityChanger.java [hunk 7]
```diff
@@ -0,0 +1,147 @@
+package com.client.core.base.tools.entitychanger.impl;
```
Comment (scriptom) [inline src/main/java/com/client/core/base/tools/entitychanger/impl/StandardEntityChanger.java:1]: how did this diff get into your PR? This is not aligned with the latest version of my PR. Can you undo this diff?

File: src/main/java/com/client/core/base/tools/propertyeditors/CustomDateTimeEditor.java [hunk 8]
```diff
@@ -8,21 +8,29 @@
...
+ * @author magnus.palm
+ *
+ */
 public class CustomDateTimeEditor extends PropertyEditorSupport {
 
     private final DateTimeFormatter customDateTimeFormatter;
 
     private final DateTimeFormatter iso8601DateTimeFormatter;
 
+    private final static String iso8601Format = "yyyy-mm-dd'T'HH:MM:SS.SSS";
```
Comment (scriptom) [inline src/main/java/com/client/core/base/tools/propertyeditors/CustomDateTimeEditor.java:23]: What's the purpose of this field?

File: src/main/java/com/client/core/base/tools/web/impl/StandardJsonConverter.java [hunk 9]
```diff
@@ -1,25 +1,24 @@
...
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
 
 @Service
 public class StandardJsonConverter implements JsonConverter {
 
-	private static Log log = LogFactory.getLog(StandardJsonConverter.class);
+	private static final Log log = LogFactory.getLog(StandardJsonConverter.class);
```
Comment (scriptom) [inline src/main/java/com/client/core/base/tools/web/impl/StandardJsonConverter.java:21]: Use Lombok

File: src/main/java/com/client/core/base/tools/web/impl/StandardJsonConverter.java [hunk 10]
```diff
@@ -60,37 +58,39 @@ public <T> T convertJsonStringToEntity(String jsonString, Class<T> type) {
...
 	}
 
+	// TODO: We could convert this to the above function {@link #convertJsonStringToEntity}. Data is simply converted to a null value here
     @Override
     public <T, E> T convertJsonStringToEntity(String jsonString, Class<T> parameterizedType, Class<E> parameterType) {
         T entity = null;
 
         try {
+			// TODO: CAN use constructParametricType(Class<?> parametrized, Class<?>... parameterClasses) instead,
+			// TODO: but I don't even think we need this whole function anymore.
```
Comment (scriptom) [inline src/main/java/com/client/core/base/tools/web/impl/StandardJsonConverter.java:74]: Let's ask Sully what he thinks. I do feel it's useless, too.

File: src/main/java/com/client/core/base/util/TriggerUtil.java [hunk 11]
```diff
@@ -109,4 +92,9 @@ public static <E extends BullhornEntity> OneToMany<E> convertReplaceAllToEntityO
 		oneToMany.setTotal(bullhornEntities.size());
 		return oneToMany;
 	}
+
+	public static <E extends BullhornEntity> void setSpecialField(E entity, Map<String,Object> valuesChanged, String sourceField, String targetField){
```
Comment (scriptom) [inline src/main/java/com/client/core/base/util/TriggerUtil.java:96]: When is this method useful?

File: src/main/java/com/client/core/base/workflow/node/AbstractWorkflowAction.java [hunk 12]
```diff
@@ -1,33 +1,38 @@
...
 
 import java.util.Locale;
 import java.util.Map;
 import java.util.Set;
 
+// TODO: This class changed quite a bit getting rid of things. Make sure there's nothing missing from original file
 public abstract class AbstractWorkflowAction<E extends BullhornEntity, T extends Traverser> implements WorkflowAction<E, T> {
 
-    private final Logger log = Logger.getLogger(getClass());
+    private final Logger log = Logger.getLogger(getClass());  // TODO: NEW LOGGER??
```
Comment (scriptom) [inline src/main/java/com/client/core/base/workflow/node/AbstractWorkflowAction.java:23]: Use lombok

File: src/main/java/com/client/core/base/workflow/node/AbstractWorkflowAction.java [hunk 13]
```diff
@@ -1,33 +1,38 @@
...
 
     private final Map<? extends BullhornRelatedEntity, Set<String>> relatedEntityFields;
 
-    private EmailTemplateService emailTemplateService;
```
Comment (scriptom) [inline src/main/java/com/client/core/base/workflow/node/AbstractWorkflowAction.java:26]: We're missing the EmailTemplateService here

File: src/main/java/com/client/core/base/workflow/node/AbstractWorkflowAction.java [hunk 14]
```diff
@@ -41,39 +46,7 @@ protected String getMessageUsingKey(String key) throws NoSuchMessageException {
...
-
-        return this.appSettings;
-    }
-
-    protected Emailer getEmailer() {
-        if(this.emailer == null) {
-            this.emailer = AppContext.getApplicationContext().getBean(Emailer.class);
-        }
-
-        return emailer;
```
Comment (scriptom) [inline src/main/java/com/client/core/base/workflow/node/AbstractWorkflowAction.java:76]: These getters need to comeback - we have no knowledge of how many projects depend on them

File: src/main/java/com/client/core/email/MailSettings.java [hunk 15]
```diff
@@ -1,120 +1,28 @@
...
+import java.util.List;
 
 /**
  * Object used to represent settings used for emailing.  Values should be provided via a properties file and injected
  * using a Spring {@link org.springframework.beans.factory.config.PropertyPlaceholderConfigurer}, either in
  * applicationContext.xml or using {@link org.springframework.beans.factory.annotation.Value} annotations
  */
+@Data
+@Configuration
 public class MailSettings {
```
Comment (scriptom) [inline src/main/java/com/client/core/email/MailSettings.java:16]: This could read from properties using `@ConfigurationProperties(prefix = "mail")`

File: src/main/java/com/client/core/email/MailSettings.java [hunk 16]
```diff
@@ -1,120 +1,28 @@
...
-    public List<String> getRouteToWhenDisabled() {
-        return routeToWhenDisabled;
-    }
-
-    public void setRouteToWhenDisabled(String[] routeToWhenDisabled) {
-        this.routeToWhenDisabled = Lists.newArrayList(routeToWhenDisabled);
-    }
-
     @Override
     public boolean equals(Object o) {
```
Comment (scriptom) [inline src/main/java/com/client/core/email/MailSettings.java:27]: Using `@Data` already takes care of `equals` and `hashCode()`. These methods can go.

File: src/main/java/com/client/core/resttrigger/controller/AbstractRestTriggerController.java [hunk 17]
```diff
@@ -31,22 +33,24 @@
     private final Class<E> type;
     private final List<TriggerValidator<E, H, T>> triggerValidators;
 
-    protected final BullhornData bullhornData;
+    @Autowired
+    protected BullhornData bullhornData;
```
Comment (scriptom) [inline src/main/java/com/client/core/resttrigger/controller/AbstractRestTriggerController.java:37]: Why not final anymore? Can't you `@Autowire` final dependencies?

File: src/main/java/com/client/core/scheduledtasks/ScheduledEventProcessing.java [hunk 18]
```diff
@@ -56,19 +56,19 @@ public void run() {
 				List<Event> events = eventResponse.getEvents();
 				List<CustomSubscriptionEvent> filteredEvents = removeEventsThrownByApiUserAndMergeDuplicateEventsAsStandardEvents(events, eventResponse.getRequestId());
 
-				log.info("Running " + subscriptionName + "subscription subscriptionEvents = " + events.size()
+				log.info("Running " + subscriptionName + " subscription subscriptionEvents = " + events.size()
```
Comment (scriptom) [inline src/main/java/com/client/core/scheduledtasks/ScheduledEventProcessing.java:59]: Thank you for fixing this

File: src/main/java/com/client/core/security/tools/rsa/GenerateRSAKeyPair.java [hunk 19]
```diff
@@ -0,0 +1,50 @@
...
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.security.*;
+
+/**
+ * Created by kcable on 6/2/2017.
+ *
+ * Used to create the public and private keys for rsa encryption.
+ */
+public class GenerateRSAKeyPair {
```
Comment (scriptom) [inline src/main/java/com/client/core/security/tools/rsa/GenerateRSAKeyPair.java:13]: Should this be included? Makes me wonder why it wasn't added earlier.

File: src/main/java/com/client/core/security/tools/rsa/RSA.java [hunk 20]
```diff
@@ -0,0 +1,102 @@
...
+import java.security.spec.PKCS8EncodedKeySpec;
+import java.security.spec.X509EncodedKeySpec;
+
+/**
+ * Created by kcable on 6/1/2017.
+ *
+ * When running locally, change file location to "src/main/resources/rsa/publicKey"
+ */
+@Component
+public class RSA extends Encryption {
```
Comment (scriptom) [inline src/main/java/com/client/core/security/tools/rsa/RSA.java:29]: Same here - What's the purpose of this? Also it references WEB-INF, which I believe we're getting rid of?

File: src/main/java/com/client/config/JpaConfig.java [hunk 21]
```diff
@@ -0,0 +1,48 @@
...
+import org.springframework.orm.jpa.support.SharedEntityManagerBean;
+import org.springframework.orm.jpa.vendor.HibernateJpaVendorAdapter;
+import org.springframework.transaction.annotation.EnableTransactionManagement;
+
+import javax.sql.DataSource;
+
+@Configuration
+public class JpaConfig {
+    @Bean
+    public LocalContainerEntityManagerFactoryBean entityManagerFactory(ApplicationSettings appSettings, DataSource dataSource) {
```
[the remark this answers is not in the document]
Comment (taswartz) [reply src/main/java/com/client/config/JpaConfig.java:19]: @scriptom 
What is the purpose of using an interface here as the return type here? There's a long list:
FactoryBean<EntityManagerFactory>, BeanClassLoaderAware, BeanFactoryAware, BeanNameAware, InitializingBean, DisposableBean, EntityManagerFactoryInfo, PersistenceExceptionTranslator, Serializable

There are also a number of interfaces extending other interfaces here. Did you have one in mind that you thought could encapsulate future behavior, because none really pop out to me.

I guess we could extend the abstract class?
AbstractEntityManagerFactoryBean

File: src/main/java/com/client/core/base/dao/impl/StandardJpaDao.java (hunk 5 above)
Comment (taswartz) [reply src/main/java/com/client/core/base/dao/impl/StandardJpaDao.java:195]: Still trying to do some testing on this (WIP)

File: src/main/java/com/client/core/base/model/helper/AbstractHelper.java (hunk 6 above)
Comment (taswartz) [reply src/main/java/com/client/core/base/model/helper/AbstractHelper.java:26]: @scriptom 
The annoying part is that the constructor for AbstractHelper is called in AbstractTriggerHelper, and AbstractScheduledTaskHelper.

That means BullhornData needs to be injected in those constructors as well... And those constructors are called in all of the implementations of the abstract helpers.

File: src/main/java/com/client/core/base/tools/entitychanger/impl/StandardEntityChanger.java (hunk 7 above)
Comment (taswartz) [reply src/main/java/com/client/core/base/tools/entitychanger/impl/StandardEntityChanger.java:1]: We discussed in the meeting that we are just going to leave this here. Then your changes will overwrite mine

File: src/main/java/com/client/core/base/tools/propertyeditors/CustomDateTimeEditor.java (hunk 8 above)
Comment (taswartz) [reply src/main/java/com/client/core/base/tools/propertyeditors/CustomDateTimeEditor.java:23]: I have absolutely no idea how this got here

File: src/main/java/com/client/core/base/tools/web/impl/StandardJsonConverter.java (hunk 9 above)
Comment (taswartz) [reply src/main/java/com/client/core/base/tools/web/impl/StandardJsonConverter.java]: I just moved all loggers to use `@Log4j2` annotation

File: src/main/java/com/client/core/base/util/TriggerUtil.java (hunk 11 above)
Comment (taswartz) [reply src/main/java/com/client/core/base/util/TriggerUtil.java:95]: ... I have no idea where this came from. That is so weird. I'm going to remove this. Thanks for pointing it out

Comment (taswartz) [reply src/main/java/com/client/core/base/util/TriggerUtil.java:95]: Oh. I think it came from the cs-starter-template

Comment (taswartz) [reply src/main/java/com/client/core/base/util/TriggerUtil.java:95]: I'm tempted to remove it, I don't know why @cablke01 added this

File: src/main/java/com/client/core/base/util/Utility.java [hunk 22]
```diff
@@ -802,32 +773,17 @@ private static <T extends SearchEntity> void searchForAll(Class<T> type, String
...
     private static <T extends SearchEntity> void searchForAll(Class<T> type, String where, Set<String> fields, Consumer<List<T>> process, Integer start) {
-        BullhornData bullhornData = getBullhornData();
-
         SearchParams params = ParamFactory.searchParams();
         params.setStart(start);
         params.setCount(BATCH_SIZE);
         params.setSort("id");
 
-        ListWrapper<T> result = bullhornData.search(type, where, fields, params);
+        ListWrapper<T> result = BULLHORN_DATA.search(type, where, fields, params);
```
Comment (taswartz) [inline src/main/java/com/client/core/base/util/Utility.java:781]: This is so strange. Why is this changed at all? I wouldn't change something like this

File: src/main/java/com/client/core/base/workflow/node/AbstractWorkflowAction.java (hunk 14 above)
Comment (taswartz) [reply src/main/java/com/client/core/base/workflow/node/AbstractWorkflowAction.java]: I knew I left a comment on this file for a reason

File: src/main/java/com/client/core/email/MailSettings.java (hunk 15 above)
Comment (taswartz) [reply src/main/java/com/client/core/email/MailSettings.java:18]: Good catch!

File: src/main/java/com/client/config/BeanConfig.java [hunk 23]
```diff
@@ -0,0 +1,51 @@
...
+    }
+
+    @Bean
+    public BullhornData bullhornData(BullhornRestCredentials bullhornRestCredentials) {
+        return new StandardBullhornData(bullhornRestCredentials);
+    }
+
+    @Bean
+    public ClassLoaderTemplateResolver templateResolver() {
+        // TODO: Search for new template resolver to have template as a string
```
[the remark this answers is not in the document]
Comment (taswartz) [reply src/main/java/com/client/config/BeanConfig.java]: `ITemplateResolver` will be returned

File: src/main/java/com/client/core/resttrigger/controller/AbstractRestTriggerController.java (hunk 17 above)
Comment (taswartz) [reply src/main/java/com/client/core/resttrigger/controller/AbstractRestTriggerController.java:34]: You cannot `@Autowired` final member variables.
HOWEVER,
You can `@Autowired` on a class constructor and then set a final member variable.

File: src/main/java/com/client/core/security/tools/rsa/GenerateRSAKeyPair.java (hunk 19 above)
Comment (taswartz) [reply src/main/java/com/client/core/security/tools/rsa/GenerateRSAKeyPair.java:13]: This might have been something that I grabbed from `cs-starter-template` on accident

File: src/main/java/com/client/config/ScheduledEventConfig.java [hunk 24]
```diff
@@ -0,0 +1,119 @@
...
+        factoryBean.setTargetMethod("run");
+        factoryBean.setConcurrent(true);
+
+        return factoryBean;
+    }
+
+    @Bean
+    ScheduledEventProcessing standardSubscriptionScheduledEventProcessing(@Value("${standardSubscriptionName}") String subName) {
+        return new ScheduledEventProcessing(subName, bullhornData, appSettings, eventWorkflowFactory);
+    }
```
Comment (taswartz) [inline src/main/java/com/client/config/ScheduledEventConfig.java:88]: The code for the standard subscription is just duplicated in the `getCustomCronTriggers()`. I'm just going to integrate these together

File: src/main/java/com/client/config/JpaConfig.java (hunk 21 above)
[the remark this answers is not in the document]
Comment (scriptom) [reply src/main/java/com/client/config/JpaConfig.java:19]: @taswartz I thought it had a friendlier interface we could've put in its place. Let's leave it as it is now.

File: src/main/java/com/client/core/security/tools/rsa/GenerateRSAKeyPair.java (hunk 19 above)
Comment (scriptom) [reply src/main/java/com/client/core/security/tools/rsa/GenerateRSAKeyPair.java:13]: @taswartz In case you're not doing it already, can we cleanup cs-starter-template stuff that might have been incorporatred here on accident?

File: src/main/java/com/client/core/base/model/helper/AbstractHelper.java (hunk 6 above)
Comment (scriptom) [reply src/main/java/com/client/core/base/model/helper/AbstractHelper.java:25]: I see. Let's leave it like it is now instead of injecting. Payoff's not good enough.

File: src/main/java/com/client/core/security/tools/rsa/GenerateRSAKeyPair.java (hunk 19 above)
Comment (taswartz) [reply src/main/java/com/client/core/security/tools/rsa/GenerateRSAKeyPair.java]: @scriptom 
Yes! I just removed these 2 files and looking for more that might have snuck in

File: src/main/java/com/client/core/base/util/Utility.java [hunk 25]
```diff
@@ -23,34 +48,12 @@
...
-import com.bullhornsdk.data.model.parameter.SearchParams;
-import com.bullhornsdk.data.model.parameter.standard.ParamFactory;
-import com.bullhornsdk.data.model.response.list.ListWrapper;
-import com.client.core.AppContext;
-import com.google.common.collect.Lists;
+    private static final int BATCH_SIZE = 200;
 
-public class Utility {
+    @Autowired
+    private static BullhornData BULLHORN_DATA;
```
Comment (taswartz) [inline src/main/java/com/client/core/base/util/Utility.java:56]: This might be strange to actually be Autowiring a dependency for a Utility class. I might have to look into another methodology for this

**Review by scriptom — CHANGES_REQUESTED** [review]
Looking very good. I would just like us to review the MailSettings bean to see if we can make it Immutable. Everything else I can live with us pushing minor versions out.

File: src/main/resources/application.properties [hunk 26]
```diff
@@ -0,0 +1,65 @@
...
+rest.prefix.custom=/custom/resttrigger
+
+# Example custom subscriptions
+scheduledtasks.customSubscriptions.standardSubscription=0 */5 * * * ?
```
[the remark this answers is not in the document]
Comment (taswartz) [reply src/main/resources/application.properties:64]: Fair. I'm fine with leaving them commented out. I just want to leave something for people to understand how to create them when forking

[the remark this answers is not in the document]
Comment (scriptom) [reply src/main/resources/application.properties:64]: I agree with that sentiment, leaving commented serves that purpose just right! We will also add this information in the README, which we are going to update thoroughly

**Review by scriptom — APPROVED** [review]
This is looking awesome. I approve!
